cranelift/codegen/src/egraph/stores.rs

```diff
@@ -0,0 +1,270 @@
+#[derive(Clone, Copy, Debug, Default, PartialEq, Eq)]
+struct LastStores {
+    heap: MemoryState,
+    table: MemoryState,
+    vmctx: MemoryState,
+    other: MemoryState,
+}
+
+/// State of memory seen by a load.
+#[derive(Clone, Copy, Debug, PartialEq, Eq, PartialOrd, Ord, Hash)]
+pub enum MemoryState {
+    Entry,
+    Store(Inst),
+    BeforeInst(Inst),
+    AfterInst(Inst),
+    Top,
+}
+
+impl std::default::Default for MemoryState {
+    fn default() -> Self {
+        MemoryState::Top
+    }
+}
+
+impl LastStores {
+    fn entry() -> Self {
+        Self {
+            heap: MemoryState::Entry,
+            table: MemoryState::Entry,
+            vmctx: MemoryState::Entry,
+            other: MemoryState::Entry,
+        }
+    }
+
+    fn update(&mut self, func: &Function, inst: Inst) {
+        let opcode = func.dfg[inst].opcode();
+        if has_memory_fence_semantics(opcode) {
+            self.heap = MemoryState::AfterInst(inst);
+            self.table = MemoryState::AfterInst(inst);
+            self.vmctx = MemoryState::AfterInst(inst);
+            self.other = MemoryState::AfterInst(inst);
+        } else if opcode.can_store() {
+            if let Some(memflags) = func.dfg[inst].memflags() {
+                if memflags.heap() {
+                    self.heap = MemoryState::Store(inst);
+                } else if memflags.table() {
+                    self.table = MemoryState::Store(inst);
+                } else if memflags.vmctx() {
+                    self.vmctx = MemoryState::Store(inst);
+                } else {
+                    self.other = MemoryState::Store(inst);
+                }
+            } else {
+                self.heap = MemoryState::AfterInst(inst);
+                self.table = MemoryState::AfterInst(inst);
+                self.vmctx = MemoryState::AfterInst(inst);
+                self.other = MemoryState::AfterInst(inst);
+            }
+        }
+    }
+
+    fn get_load_input_state(&self, func: &Function, inst: Inst) -> MemoryState {
+        if let Some(memflags) = func.dfg[inst].memflags() {
+            if memflags.heap() {
+                self.heap
+            } else if memflags.table() {
+                self.table
+            } else if memflags.vmctx() {
+                self.vmctx
+            } else {
+                self.other
+            }
+        } else {
+            MemoryState::AfterInst(inst)
+        }
+    }
+
+    fn meet_from(&mut self, other: &LastStores, loc: Inst) {
+        let meet = |a: MemoryState, b: MemoryState| -> MemoryState {
+            match (a, b) {
+                (a, b) if a == b => a,
+                (MemoryState::Top, a) | (a, MemoryState::Top) => a,
+                _ => MemoryState::BeforeInst(loc),
+            }
+        };
+
+        self.heap = meet(self.heap, other.heap);
+        self.table = meet(self.table, other.table);
+        self.vmctx = meet(self.vmctx, other.vmctx);
+        self.other = meet(self.other, other.other);
+    }
```

Comment:
If you get rid of `MemoryState::Top` it becomes pretty straightforward to have `meet_from` return whether any of the states changed, which could simplify `compute_block_input_states`:

```suggestion
    fn update(&mut self, other: &LastStores, loc: Inst) -> bool {
        let meet = |a: &mut MemoryState, b: MemoryState| -> MemoryState {
            if *a == b {
                false
            } else {
                *a = MemoryState::BeforeInst(loc);
                true
            }
        };

        meet(&mut self.heap, other.heap)
            | meet(&mut self.table, other.table)
            | meet(&mut self.vmctx, other.vmctx)
            | meet(&mut self.other, other.other)
    }
```

Reply:
The second version reads more confusingly to me; and it seems less error-prone to retain the top-level "if old != new" check than to thread a bool through the recomputation. If you really feel strongly here I'm happy to tweak further though...

Reply:
Nah, I don't feel strongly about this.